### PR TITLE
Geo variogram

### PR DESCRIPTION
**Pull request by LSchueler**
A first step towards supporting spherical coordinate systems #54 is to support variogram estimation on spherical coordinate systems.

This is a first proposal for calculating unstructured variograms on spheres.

A few things to discuss:

- [x] is `spherical` a good name for the argument in the function `vario_estimate_unstructured`?
  - let's use `great-circle`
- [ ] how are we going to provide the new distance function to users? - This is important in order to, e.g. create bins.
  - Possible options:
    - import the Cython function from `estimator.pyx` into one of the `tools.py` namespaces
    - create a new `tools.pyx` as a more general Cython code collection
    - reimplement the distance function in Python
    - use on of the already existing Python packages as external dependenies, e.g. [haversine](https://github.com/mapado/haversine)
  - Suggestion:
    - [ ] users shouldn't need to care about the haversine routines (just use it internally)
    - [ ] the rest of GSTools should deal with xyz positions converted by a new routine `latlon2xyz`
    - [ ] variogram fitting needs to provide a new workflow to fit associated 3D *Yadrenko* models to estimated great-circle variograms
    - [ ] with all this, we don't need to rewrite the `SRF`, `CovModel` and `Krige` classes
    - [ ] we need a set of good tutorials dealing with geo-coordinates (explaining yadrenko models and great-circle distances)
- [ ] better testing: I'm not sure what else to test
- [ ] how could potential pitfalls / confusing arguments be caught (lat-lon, y-x)
  - as written above: conversions routines and good tutorials
- [ ] should we stick to the unit of metre or does kilometre make more sense?
  - we shoud use the great-circle distance in `deg` and leave this decision to the user

### CODE REVIEW

File: gstools/variogram/estimator.pyx [hunk 1]
```diff
@@ -36,6 +36,27 @@ cdef inline double _distance_2d(
...
+    cdef double diff_lat = (x[j] - x[i]) * deg_2_rad
+    cdef double diff_lon = (y[j] - y[i]) * deg_2_rad
+    cdef double arg = (
+        pow(sin(diff_lat/2.0), 2) +
+        cos(x[i]*deg_2_rad) *
+        cos(x[j]*deg_2_rad) *
+        pow(sin(diff_lon/2.0), 2)
+    )
+    # earth radius for WGS84 ellipsoid r ~ 6371km
+    return 12742000.0 * atan2(sqrt(arg), sqrt(1.0-arg))
```
Comment (MuellerSeb) [inline gstools/variogram/estimator.pyx:58]: I would use the great-circle distance in `deg` as the result, so we don't have to deal with the earth radius. This is also done in Pykrige this way ATM. Then we also don't have to care about `km` vs `m` and leave this decision to the user.

File: gstools/variogram/variogram.py [hunk 2]
```diff
@@ -32,6 +32,18 @@ def _set_estimator(estimator):
         )
     return cython_estimator
 
+def _set_distance(distance):
+    """Translate the verbose Python distance identifier to single char."""
+    if distance.lower() == "euclidean":
+        cython_distance = "e"
+    elif distance.lower() == "spherical":
```
Comment (MuellerSeb) [inline gstools/variogram/variogram.py:39]: Same argument: Let's use `"great-circle"` here.

File: gstools/variogram/variogram.py [hunk 3]
```diff
@@ -92,6 +105,13 @@ def vario_estimate_unstructured(
             * "cressie": an estimator more robust to outliers
 
         Default: "matheron"
+    distance : :class:`str`, optional
+        the distance function, possible choices:
+
+            * "euclidean": the distance in Euclidean spaces
+            * "spherical": the great-circle distance on spheres
```
Comment (MuellerSeb) [inline gstools/variogram/variogram.py:112]: Same argument: Let's use "great-circle" here.

File: gstools/variogram/variogram.py (hunk 2 above)
Comment (LSchueler) [reply gstools/variogram/variogram.py]: Good idea

File: gstools/variogram/estimator.pyx (hunk 1 above)
Comment (MuellerSeb) [reply gstools/variogram/estimator.pyx:58]: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.haversine_distances.html

sklearn takes lat-lon in deg and returns to distance in rad. We could use that as a guide.

**Comment by MuellerSeb** [comment]
this is now made obsolete by #113

**Pull request closed without merging**